### PR TITLE
Mitigation for current tool test skips on CI

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Mitigation for https://github.com/flutter/flutter/issues/46142, ensures that _most_ tests are being run, but not confirmed yet.

Currently the commands.shard appears to only be running tests for a single subdirectory (due to https://github.com/dart-lang/test/issues/1116), permeable/, while the hermetic/ subdirectory is entirely skipped. To work around this a bit, when selecting the commands.shard I invoke pub twice, once in each sub directory. Unfortunately this has revealed that several of the tests were failing, and further more that shard crashes on windows (https://github.com/flutter/flutter/issues/46180). 

I also attempt to solve this by adjusting the build_runner configuration, but found that the failing tests did not cause a non-zero exit code to be set: https://github.com/dart-lang/build/issues/2562

Due to other bugs in pub/test, we might be skipping tests elsewhere - thus this is only a partial mitigation.

EDIT:

Thanks to @grouma that tracked the issue down. Ultimately the problem was a test calling io.exit which takes down the entire test runner.

### CODE REVIEW

**Comment by zanderso** [comment]
Could you expand the PR description a bit to explain what the mitigations are and cross reference with other issues?

File: dev/bots/test.dart [hunk 1]
```diff
@@ -260,6 +262,30 @@ Future<void> _runToolTests() async {
       .map<String>((String name) => path.basenameWithoutExtension(name)),
     // The `dynamic` on the next line is because Map.fromIterable isn't generic.
     value: (dynamic subshard) => () async {
+      if (subshard == 'commands') {
```
Comment (zanderso) [inline dev/bots/test.dart]: Do the subdirectories of test/general.shard/ need this treatment? If so, I think a more manual walk of the test directory tree would be preferable here.

Comment (jonahwilliams) [reply dev/bots/test.dart]: Surprisingly, it seems like pub is running these tests as expected

Comment (zanderso) [reply dev/bots/test.dart]: On further thought, I'm wondering about the wisdom of trusting it to do that before https://github.com/dart-lang/test/issues/1116 is resolved.

Comment (jonahwilliams) [reply dev/bots/test.dart]: Agreed, but I still think this change is an improvement overall.

I also experimented with just passing every test file individually, but that hit the terminal line limit on windows and still seemed to be skipping tests.

**Review by zanderso — APPROVED** [review]

**Comment by jonahwilliams** [comment]
Updating with fixes to upgrade test

File: packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart [hunk 2]
```diff
@@ -238,16 +243,17 @@ void main() {
 
       testUsingContext('upgrade continue prints welcome message', () async {
         final UpgradeCommand upgradeCommand = UpgradeCommand(fakeCommandRunner);
-        await runner.run(
+        applyMocksToCommand(upgradeCommand);
+
+        await createTestCommandRunner(upgradeCommand).run(
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart]: Without the test command runner, this was actually invoking io.exit

Comment (zanderso) [reply packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart]: Great catch!

Comment (jonahwilliams) [reply packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart]: 99% @grouma

**Pull request merged**